dart/lib/src/client_reports/client_report_recorder.dart

```diff
@@ -0,0 +1,53 @@
+import 'package:meta/meta.dart';
+
+import '../utils.dart';
+import 'client_report.dart';
+import 'discarded_event.dart';
+import 'outcome.dart';
+import '../transport/rate_limit_category.dart';
+
+@internal
+class ClientReportRecorder {
+  ClientReportRecorder([this._dateTimeProvider = getUtcDateTime]);
+
+  final DateTime Function() _dateTimeProvider;
+  final Map<_QuantityKey, int> _quantities = {};
+
+  void recordLostEvent(final Outcome reason, final RateLimitCategory category) {
+    final key = _QuantityKey(reason, category);
+    var current = _quantities[key] ?? 0;
+    _quantities[key] = current + 1;
+  }
+
+  ClientReport? flush() {
+    if (_quantities.isEmpty) {
+      return null;
+    }
+
+    final events = _quantities.keys.map((key) {
+      final quantity = _quantities[key] ?? 0;
+      return DiscardedEvent(key.reason, key.category, quantity);
+    }).toList();
+
+    _quantities.clear();
+
+    return ClientReport(_dateTimeProvider(), events);
+  }
+}
+
+class _QuantityKey {
+  _QuantityKey(this.reason, this.category);
+
+  final Outcome reason;
+  final RateLimitCategory category;
+
+  @override
+  int get hashCode => Object.hash(reason, category);
+
+  @override
+  bool operator ==(dynamic other) {
+    return other is _QuantityKey &&
+        other.reason == reason &&
+        other.category == category;
+  }
```

Comment:
We usually don't need `hashCode` and `operator`, why do we need them here?

Reply:
We are using the combination of reason and category as a map key, so we can count all combinations of both. JavaScript also does this, but uses a concatenated string of both. To get to the correct typed objects again from a string key, we'd need to split and also implement code to get to the correct enum cases. It's way more convenient to use it this way and also it's a file private class here.

Reply:
If there are benefits during the code execution, we can keep it, we just don't want it if it's for testing only, the idea is to keep the SDK as light as possible.

Reply:
It's not for testing only, it's an important implementation detail of the component. Events with the same outcome but different categories are counted separately.

Reply:
Thanks